### ISSUE TITLE
Add type filter in smartcontent

### ISSUE THREAD

**Issue by alexander-schranz** [sulu/sulu#1456]
E.g Content Structure
- news (overview template)
  - 2014 (overview template)
    - july (overview template)
      - Page 1 (news template)
      - Page 2 (news template)
  - 2015 (overview template)
    - august (overview template)
      - Page 1 (news template)
      - Page 2 (news template)
    - september (overview template)
      - Page 1 (news template)
      - Page 2 (news template)

I want to have a smartcontent for a specific template. Because I have nested overviews and I don't want that the overview templates are shown in the smartcontent.

**Comment by danrot** [comment]
I've swapped template with type in the title. Would prefer that naming, because then it is easier to be reused in other contexts as well.

**Comment by alexander-schranz** [comment]
Would think about the following in the Smart Content DataProvider [Configuration](https://github.com/sulu/sulu/blob/2be30ce7561911b243884ae27e86907899fa2639/src/Sulu/Component/Content/SmartContent/PageDataProvider.php#L112-L133):

```xml
    ->enableTemplates('page'); // the pages template "type" key
```

And for articles:

```php
    ->enableTemplates('article'); // the articles template "type" key
```

**Comment by danrot** [comment]
In order to not couple this to pages and articles, I would prefer to call this function `enableTypes`. Maybe we want to filter for only a specific type of contacts in the future.

And I think that is something that should be exposed to the content manager. Then a call like the following should be sufficient:

```php
->enableTypes()
```

**Comment by alexander-schranz** [comment]
@danrot enableTypes sounds good for me as it are template types. At current state I would just implement it for templates as there should it be as simple as possible. For contacts or other types we would need some kind of providers or resources for that types but think in first case I would just go with template types.

**Comment by danrot** [comment]
When you e.g. load the metadata for the `pages` resource key, you already get all available types. I would just reuse that information, should work for other entities in a similar fashion.

**Comment by alexander-schranz** [comment]
@danrot thats what I would implement currently, but for other entities this types do not exist as they don't have different template types. Media just have media types which would make sense in the smart content filter but they have nothing todo with form metadata types that is what I meant with it that there you would need something else.

**Comment by danrot** [comment]
Ok, thought of this types as something different :see_no_evil: We could make the parameter in `enableTypes` optional, and if nothing is passed it tries to get that information from the metadata. Optionally the available types can be passed, for cases like media.

**Comment by danrot** [comment]
Fixed in #5649

**Comment by danrot** [comment]
We have only implemented the types one level deep, and this will be configured like this:

```php
protected function configure(BuilderInterface $builder): void
    {
        $builder
            ->enableTags()
            ->...()
            ->enableTypes(
                [
                    ['type' => 'homepage', 'title' => 'my-translation-key'],
                    ['type' => 'default', 'title' => ['de' => 'Standard', 'en' => 'Default']],
                ]
            );
    }
```

E.g. for articles we have said that two levels might make sense, because there are types and templates. For that usecase the above example could be extended like this:

```php
    protected function configure(BuilderInterface $builder): void
    {
        $builder
            ->enableTags()
            ->...()
            ->enableTypes(
                [
                    [
                        'type' => 'first-article-type', 
                        'title' => 'first-translation-key',
                        'children' => [
                            ['type' => 'first-article-type', 'title' => 'first-translation-key'],
                        ],
                    ],
                    ['type' => 'last-artice-type', 'title' => 'second-translation-key'],
                ]
            );
    }
```
